### PR TITLE
feat(core): add option for `require_literal_leading_dot`, closes #6158

### LINKED ISSUE #6158: [bug] `assetScope` suddenly no longer working?
### Describe the bug

I tried upgrading my tauri versions (from 1.2.0) and suddenly can't load any assets. For example, when loading an asset now I get these errors:

```
asset protocol not configured to allow the path: /home/ftseng/.config/hundun/.thumbs/references.media.environments.coast.rzbu4ds397p91.jpg.jpg
```

and in the frontend:

```
[Error] Unhandled Promise Rejection: path not allowed on the configured scope: /home/ftseng/.config/hundun/.thumbs
```

I've tried a number of different `assetScope` configs, even trying the complete paths for some of the assets, to no avail:

```
 "assetScope": ["$APP/*", "$APPCONFIG/*", "$APPDATA/*", "*", "/home/ftseng/.config/hundun/.thumbs/references.media.environments.coast.rzbu4ds397p91.jpg.jpg"]
```

and

```
"assetScope": ["$APP/*", "$APPCONFIG/*", "$APPDATA/*", "*",
            "$APP/**", "$APPCONFIG/**", "$APPDATA/**"]
```

and

```
"assetScope": ["**"]
```

This is the url for the example that is being passed to the `img` tag: ` "asset://localhost/%2Fhome%2Fftseng%2F.config%2Fhundun%2F.thumbs%2Freferences.media.environments.coast.rzbu4ds397p91.jpg.jpg"`, produced using `convertFileSrc`.

I seem to be having a similar issue with the `fs.scope` config, where I can't iterate over files anymore.

Am I just misconfiguring something? I'm totally stumped.

### Reproduction

1. Try to load the asset in HTML

### Expected behavior

The assets to load without issue

### Platform and versions

> hundun@0.1.0 tauri
> tauri "info"

Environment
  › OS: Ubuntu 22.04 X64
  › Node.js: 18.1.0
  › npm: 8.8.0
  › pnpm: Not installed!
  › yarn: Not installed!
  › rustup: 1.25.1
  › rustc: 1.68.0-nightly
  › cargo: 1.68.0-nightly
  › Rust toolchain: nightly-x86_64-unknown-linux-gnu 

Packages
  › @tauri-apps/cli [NPM]: 1.2.3
  › @tauri-apps/api [NPM]: 1.2.0
  › tauri [RUST]: 1.2.4,
  › tauri-build [RUST]: 1.2.1,
  › tao [RUST]: 0.15.8,
  › wry [RUST]: 0.23.4,

App
  › build-type: bundle
  › CSP: default-src 'self'; img-src 'self' asset: https://asset.localhost
  › distDir: ../build
  › devPath: http://localhost:8080/
  › framework: React
  › bundler: Webpack

App directory structure
  ├─ tauri
  ├─ assets
  ├─ .git
  ├─ public
  ├─ src
  ├─ node_modules
  └─ scripts

### Stack trace

_No response_

### Additional context

My complete config:

```
{
  "$schema": "../node_modules/@tauri-apps/cli/schema.json",
  "build": {
    "beforeBuildCommand": "npm run build",
    "beforeDevCommand": "npm run start",
    "devPath": "http://localhost:8080",
    "distDir": "../build"
  },
  "package": {
    "productName": "hundun",
    "version": "0.1.0"
  },
  "tauri": {
    "allowlist": {
      "shell": {
        "open": true
      },
      "dialog": {
        "all": true
      },
      "protocol": {
        "asset": true,
        "assetScope": ["**", "**/*"]
      },
      "fs": {
        "all": true,
        "scope": ["**", "**/*"]
      },
      "path": {
        "all": true
      }
    },
    "bundle": {
      "active": true,
      "category": "DeveloperTool",
      "copyright": "",
      "deb": {
        "depends": []
      },
      "externalBin": [],
      "icon": [
        "icons/32x32.png",
        "icons/128x128.png",
        "icons/128x128@2x.png",
        "icons/icon.icns",
        "icons/icon.ico"
      ],
      "identifier": "hundun",
      "longDescription": "",
      "macOS": {
        "entitlements": null,
        "exceptionDomain": "",
        "frameworks": [],
        "providerShortName": null,
        "signingIdentity": null
      },
      "resources": [],
      "shortDescription": "",
      "targets": "all",
      "windows": {
        "certificateThumbprint": null,
        "digestAlgorithm": "sha256",
        "timestampUrl": ""
      }
    },
    "security": {
      "csp": "default-src 'self'; img-src 'self' asset: https://asset.localhost"
    },
    "updater": {
      "active": false
    },
    "windows": [
      {
        "fullscreen": false,
        "height": 600,
        "resizable": true,
        "title": "hundun",
        "width": 800
      }
    ]
  }
}
```

### PR DESCRIPTION
**Pull request by amrbashir**
### What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Docs
- [ ] New Binding issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

### Does this PR introduce a breaking change?

- [ ] Yes, and the changes were approved in issue #___
- [x] No

### Checklist
- [x] When resolving issues, they are referenced in the PR's title (e.g `fix: remove a typo, closes #___, #___`)
- [x] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).
- [x] I have added a convincing reason for adding this feature, if necessary

### CODE REVIEW

File: core/tauri-config-schema/schema.json [hunk 1]
```diff
@@ -2069,6 +2069,13 @@
               "items": {
                 "type": "string"
               }
+            },
+            "requireLiteralLeadingDot": {
+              "description": "Whether or not paths that contain components that start with a `.` will require that `.` appears literally in the pattern; `*`, `?`, `**`, or `[...]` will not match. This is useful because such files are conventionally considered hidden on Unix systems and it might be desirable to skip them when listing files.\n\nDefaults to `true` on Unix systems and `true` on Windows",
```
Comment (simonhyll) [inline core/tauri-config-schema/schema.json]: I think you meant to say it's false on Unix.
```suggestion
              "description": "Whether or not paths that contain components that start with a `.` will require that `.` appears literally in the pattern; `*`, `?`, `**`, or `[...]` will not match. This is useful because such files are conventionally considered hidden on Unix systems and it might be desirable to skip them when listing files.\n\nDefaults to `false` on Unix systems and `true` on Windows",
```

File: tooling/cli/schema.json [hunk 2]
```diff
@@ -2069,6 +2069,13 @@
               "items": {
                 "type": "string"
               }
+            },
+            "requireLiteralLeadingDot": {
+              "description": "Whether or not paths that contain components that start with a `.` will require that `.` appears literally in the pattern; `*`, `?`, `**`, or `[...]` will not match. This is useful because such files are conventionally considered hidden on Unix systems and it might be desirable to skip them when listing files.\n\nDefaults to `true` on Unix systems and `true` on Windows",
```
Comment (simonhyll) [inline tooling/cli/schema.json]: Same again, I think you meant to say it's false on Unix.
```suggestion
              "description": "Whether or not paths that contain components that start with a `.` will require that `.` appears literally in the pattern; `*`, `?`, `**`, or `[...]` will not match. This is useful because such files are conventionally considered hidden on Unix systems and it might be desirable to skip them when listing files.\n\nDefaults to `false` on Unix systems and `true` on Windows",
```

File: core/tauri-utils/src/config.rs [hunk 3]
```diff
@@ -1316,6 +1291,16 @@ pub enum FsAllowlistScope {
     /// This gets precedence over the [`Self::Scope::allow`] list.
     #[serde(default)]
     deny: Vec<PathBuf>,
+    /// Whether or not paths that contain components that start with a `.`
+    /// will require that `.` appears literally in the pattern; `*`, `?`, `**`,
+    /// or `[...]` will not match. This is useful because such files are
+    /// conventionally considered hidden on Unix systems and it might be
+    /// desirable to skip them when listing files.
+    ///
+    /// Defaults to `true` on Unix systems and `true` on Windows
```
Comment (simonhyll) [inline core/tauri-utils/src/config.rs]: It's false on Unix by default.
```suggestion
    /// Defaults to `false` on Unix systems and `true` on Windows
```

**Review by simonhyll — CHANGES_REQUESTED** [review]
Still have to actually test the code but it looks good to me thus far.

File: core/tauri-config-schema/schema.json (hunk 1 above)
Comment (amrbashir) [reply core/tauri-config-schema/schema.json]: yeah, thanks for catching it

**Review by simonhyll — APPROVED** [review]
I have verified the behavior now running in WSL and it seems to function correctly.

**Review by wusyong — APPROVED** [review]

File: core/tauri-utils/src/config.rs [hunk 4]
```diff
@@ -38,6 +38,10 @@ use crate::TitleBarStyle;
 
 pub use self::parse::parse;
 
+fn default_true() -> bool {
```
Comment (tweidinger) [inline core/tauri-utils/src/config.rs:41]: Shoudln't this be moved into another PR as it is not required for the added option and distinct change?

File: core/tauri/src/scope/fs.rs [hunk 5]
```diff
@@ -106,10 +107,29 @@ impl Scope {
       }
     }
 
+    let require_literal_leading_dot = match scope {
+      FsAllowlistScope::Scope {
+        require_literal_leading_dot: Some(require),
+        ..
+      } => *require,
+      // dotfiles are not supposed to be exposed by default on unix
+      #[cfg(unix)]
```
Comment (tweidinger) [inline core/tauri/src/scope/fs.rs:116]: Actually reversed behavior and should be true on unix by default.

When `require_literal_leading_dot` is set to `true` it will strip the file paths, which contain `.` in their first path component (eg. `.ssh`).
By setting `require_literal_leading_dot` to false by default it will return the *hidden* folders and files on unix.

File: core/tauri/src/scope/fs.rs [hunk 6]
```diff
@@ -271,6 +281,17 @@ mod tests {
       allowed_patterns: Default::default(),
       forbidden_patterns: Default::default(),
       event_listeners: Default::default(),
+      match_options: glob::MatchOptions {
+        // this is needed so `/dir/*` doesn't match files within subdirectories such as `/dir/subdir/file.txt`
+        // see: https://github.com/tauri-apps/tauri/security/advisories/GHSA-6mv3-wm7j-h4w5
+        require_literal_separator: true,
+        // dotfiles are not supposed to be exposed by default on unix
+        #[cfg(unix)]
+        require_literal_leading_dot: false,
```
Comment (tweidinger) [inline core/tauri/src/scope/fs.rs:290]: Same as for above

File: core/tauri-utils/src/config.rs (hunk 4 above)
Comment (amrbashir) [reply core/tauri-utils/src/config.rs:41]: yeah it should've, will take this into consideration and try to keep the PRs minimal

File: core/tauri/src/scope/fs.rs (hunk 5 above)
Comment (amrbashir) [reply core/tauri/src/scope/fs.rs:116]: damn me 😅, I had the logic flipped inside my head the whole time I was working on the PR.
